### DISCUSSION TITLE
WhenAnyValue might not send latest value

### DISCUSSION THREAD

**Discussion by erikm-g** [reactiveui/ReactiveUI#3189] [Q&A] [answered]
**Describe the bug**
If a subscription is started for changes on a property using WhenAnyValue and the property is updated on another thread at approximately the same time, WhenAnyValue might send the old value as the initial value in the observable sequence and not send the updated value at all. 

**Steps To Reproduce**
Let one thread update a property and set up a subscription on changes for it from another. This unit test illustrates the problem. It passes most of the times, but if repeated often enough it will eventually fail. The value of anyReactiveObject.Value will be 2 but no call to the subscription is emitted.
´´´
    private class SomeReactiveObject : ReactiveObject
    {
        private int value;

        public int Value
        {
            get => value;
            set => this.RaiseAndSetIfChanged(ref this.value, value);
        }
    }

    [Test]
    [Repeat(10000)]
    public async Task WhenAnyShouldBeThreadSafe()
    {
        var anyReactiveObject = new SomeReactiveObject { Value = 1 };
        var task = Task.Run(() => anyReactiveObject.Value = 2);
        bool found = false;
        anyReactiveObject.WhenAnyValue(x => x.Value)
            .Where(v => v == 2).Subscribe(_ => found = true);
        await task;
        Assert.That(found);
    }
´´´
**Expected behaviour**
The test above should pass.
In a multithreaded environment, events may occur out of order, or if updates are very frequent some intermediate values might be missed, but the latest value should always be emitted to the pipeline.

**Environment**
- OS:  Windows 10
- Device:  PC
- ReactiveUI Version:  16.2.6

**Comment by kmgallahan** [comment]
Isn't this just a standard race condition?

Both threads are running in parallel and nothing is stopping the subscription from being setup after the Task has completed.

> WhenAnyValue might send the old value as the initial value in the observable sequence

No initial value is "sent", you just Assert that found = false at the end, which is what you initialize it to. The subscription is never called if it is setup after the value is changed.

**Reply by kmgallahan** [reply]
@RolandPheasant I wouldn't mark this as the answer. My comment:

> The subscription is never called if it is setup after the value is changed.

Is false and was based on my misunderstanding of WhenAnyValue at the time - I was unaware it emits this initial value (except when observing a property chain that throws a null exception).

Disregarding the fact that `bool` isn't thread safe and therefore is in an indeterministic state for the assertion (I think?), a more accurate answer is:

> WhenAnyValue is not guaranteed to emit the latest value of a property if setup in a multithreaded environment.

**Reply by glennawatson** [reply]
Side note Bool are atomic read, writes but not read and write at the same time. 

They are thread safe for read or write operations but most people want atomic combined read and write. 

I've used the interlocked.exchange with a numeric to achieve a gate in past iterations.

**Comment by erikm-g** [comment]
If the task completes before the subscription is setup, WhenAnyValue emits 2 and the test passes.
If the subscription is setup before the task starts to execute WhenAnyValue will emit first 1 and then 2 and the test passes.
If the task is executed at the same time as the subscription is setup, sometimes WhenAnyValue emits only the value 1 but the property's value is 2. This causes the test to fail.

**Comment by erikm-g** [comment]
If I would modify the test and add .Do(TestContext.Write) before the .Where, I would expect the test to sometimes print 12, sometimes print just 2 and in a few occasions print 21, but in all cases I would expect the test to pass and never to print just 1 and fail. There is no race condition between assigning the values 1 and 2.

**Comment by glennawatson** [comment]
There's a race condition in your Code. The task will start immediately which may or may not be before the WhenAnyValue is registered. 

Try converting your task into a Func<Task> and making it into a function you call after the WhenAnyValue is registered and then await that.

**Comment by kmgallahan** [comment]
@glennawatson I tested this code:

```cs
public MainWindowViewModel()
{
    var i = 1000;
    do
    {
        i--;
        WhenAnyShouldBeThreadSafe();
    } while (i > 0);
}

public void WhenAnyShouldBeThreadSafe()
{
    var anyReactiveObject = new SomeReactiveObject { Value = 1 };
    Task.Run(() => anyReactiveObject.Value = 2);
    anyReactiveObject.WhenAnyValue(x => x.Value)
        .Subscribe(x =>
        {
            if (anyReactiveObject.Value is 1) Debug.WriteLine(anyReactiveObject.Value);
        });
    }
}

public class SomeReactiveObject : ReactiveObject
{
    public int Value
    {
        get => value;
        set => this.RaiseAndSetIfChanged(ref this.value, value);
    }

    private int value;
}
```

In theory `1` should never be written to debug out because the subscription is always made after ` var anyReactiveObject = new SomeReactiveObject { Value = 1 };`. However, in the 1000 runs it does show up multiple times for me...

**Comment by glennawatson** [comment]
`Task.Run(() => anyReactiveObject.Value = 2); // This can run or not run before the WhenAnyValue happens, use a Func<Task>`
    anyReactiveObject.WhenAnyValue(x => x.Value)
        .Subscribe(x =>
        {
            if (anyReactiveObject.Value is 1) Debug.WriteLine(anyReactiveObject.Value);
        });

**Comment by kmgallahan** [comment]
@glennawatson Without `Task.Run(() => anyReactiveObject.Value = 2);` the observable / subscription should never fire because the subscription is made after the value is assigned here `var anyReactiveObject = new SomeReactiveObject { Value = 1 };`, correct?

That would mean Value = 1 should never show up in the observable / subscription (be sent to debug out), or am I missing something?

**Comment by glennawatson** [comment]
`Func<Task> myFancyValueChanger = () => Task.Run(() => .... `

`something.WhenAnyValue(x => ....`
`await myFancyValueChanger.Invoke();`

**Comment by kmgallahan** [comment]
@glennawatson 

Ignoring all of the code above - from my understanding, just these 2 lines of code should never fire the subscription, correct?

```cs
anyReactiveObject.Value = 1;
anyReactiveObject.WhenAnyValue(x => x.Value).Subscribe(x => Debug.WriteLine(anyReactiveObject.Value));
```

So value = 1 should **never** show up in the debug output.

This code however will very seldomly show 1 in the debug output:

```cs
anyReactiveObject.Value = 1;
Task.Run(() => anyReactiveObject.Value = 2);
anyReactiveObject.WhenAnyValue(x => x.Value).Subscribe(x => Debug.WriteLine(anyReactiveObject.Value));
```

I can't reason as to why 1 should ever show up, but it does.

**Comment by glennawatson** [comment]
You don't control when Task.Run runs, it's up to the task scheduler

**Comment by kmgallahan** [comment]
@glennawatson 

I understand there is a race condition happening.

In this race condition there should be two scenarios:

1. Value is set to 2 **AFTER** the subscription is setup, so the subscription is called printing 2 to the output
2. Value is set to 2 **BEFORE** the subscription is setup, so nothing is printed to the output

There should be no scenario 3:

3. Value is set to 2 **AFTER** the subscription is setup, the subscription prints **1** to the output

But that 3rd scenario does indeed occur 5-10 times out of 1000.

@RLittlesII Can you sanity check me here?

**Comment by glennawatson** [comment]
`WhenAnyValue` always emits the current initial value of the Property unless you do a Skip(1)

**Comment by glennawatson** [comment]
Scenario 3 can happen

1. Initial Value is 1
2. Task sets value to 2
3. WhenAnyValue grabs initial value of property and emits 2

**Comment by RLittlesII** [comment]
@kmgallahan 
> There should be no scenario 3:

> Value is set to 2 AFTER the subscription is setup, the subscription prints 1 to the output
But that 3rd scenario does indeed occur 5-10 times out of 1000.

The fact you are using `Task.Run` by definition means you are not in control of **when** it is set.  Asynchronous code does not happen where you write the line of code.  It happens at some point in the future.

**Comment by kmgallahan** [comment]
Sorry guys, didn't know it emitted the initial value on setup, please forget I ever existed, lol

I checked the docs [here](https://www.reactiveui.net/docs/handbook/when-any/#whenanyvalue) when I first saw this issue and it only mentions things like this for a while:

> This returns an observable that yields the current value of Foo each time it changes:

> You can Subscribe to the observable returned by the WhenAny variant and get notifications whenever the value changes.

Funny thing is I've been using it for months now...

**Comment by glennawatson** [comment]
I was just reading the docs so I could link you to them and thought the fact they emit the initial value wasn't very well described. If I get time I'll make it more obvious today.

**Comment by glennawatson** [comment]
FYI https://github.com/reactiveui/website/pull/745

**Comment by kmgallahan** [comment]
@glennawatson 

Thanks for the quick update.

I remember now why I assumed the initial value wasn't emitted for months - the [IntelliSense descriptions](https://github.com/reactiveui/ReactiveUI/blob/main/src/ReactiveUI/VariadicTemplates.cs) read:

> WhenAnyValue allows you to observe whenever the value of a property on an object has changed, providing an initial value when the Observable is set up

Yet during my first tests that seemed to not be happening, so I just thought: ¯\\_(ツ)\_/¯ I guess the docs are wrong.

Now I realize I must have only been testing it where the initial values were null (but only for chained properties), which is a significant gotcha.

I'd suggest that text be updated as well.

```cs
a = null; // emits initial value
a.z = null; // does not emit initial value
b = object; // emits initial value
b.y = null; // emits initial value
```

**Comment by erikm-g** [comment]
@glennawatson There is no race condition between setting the value to 1 and 2. There is a race condition between setting the value to 2 and setting up the subscription; that's the whole point of the test. To make the simplest possible code where some property is updated on another thread at the same time as a subscription is started. If the update occurs first, 2 should be emitted as the initial value. If the subscription is set up first 1 should be emitted as the initial value and 2 as an updated value. What causes the test to fail is that 2 is not emitted at all. If you close this bug saying that this is expected behavior you should really update the documentation to say that reactive UI should not be used for multi threaded applications.

**Comment by kmgallahan** [comment]
@erikm-g WhenAnyValue emits the initial value when you call Subscribe on it (except for a chain of properties where an accessor is used on a null value).

Please read the above last few comments.

**Comment by erikm-g** [comment]
Yes, i know. Which is why 2 should always be emitted since it is always the final value. But it is not.

**Comment by kmgallahan** [comment]
I think the problem is that you misunderstand how `await` and `Task.Run` works.

You saying `await task` has no impact whatsoever on when the task is run, because you ran the Task when you called `Task.Run` already.

**Comment by erikm-g** [comment]
It is actually 5 lines of code:
 1. Set Value = 1 on Thread A
 2. Run Task to set Value = 2 on a Thread B
 3. Subscribe on Thread A and check if value is 2
 4. Await the task started in step 2 so that we know that the value has been set to 2 before continuing to step 5. Since no other code is updating the value, it must be equal to 2 now.
 5. Assert to check if Value has ever been 2. Assert fails (sometimes)

I don't care about the order things are happening and I don't care if I ever receive the value 1. I am subscribing to the initial value and every future updates on it, the subscription is still active, the value is currently 2 and the PropertyChanged event handler has finished executing, so I expect to have received the value 2 at some point on some thread.

**Comment by erikm-g** [comment]
I know how await and Task.Run works. When await task returns I know that the task has completed, that is what is important here. I want the task and setting up the subscription to be run in parallel in this test since that is what is triggering the error. In my real application it is a completely different part of the code running on a different thread updating the value. This is just a very simple example of reproducing the behavior of a multi threaded application where my WhenAnyValue subscriptions does not trigger if the value is updated at the same time as the subscription is set up.

**Comment by kmgallahan** [comment]
I see your problem now. The subscription is run on the thread that changed the value, not the one where it was setup.

When the Task runs on a different thread then the bool is changed on that thread, and bool is inherently not thread safe, so back on the original thread it may or may not Assert to true at the end.

If you make sure to change the bool on the original thread in the subscription then it should work as you expect.

**Comment by erikm-g** [comment]
That is not the issue. The problem is that the value 2 is never emitted. I can change the bool to an int, increment it with Interlocked.Increment in the subscription and add a memory barrier before my assert, it is still the same error. The lambda passed to subscribe is simply never executed.

**Comment by kmgallahan** [comment]
> The problem is that the value 2 is never emitted.
> The lambda passed to subscribe is simply never executed.

I have no idea what you are talking about now. If the lamba didn't execute then why does this print to debug output?

```cs
anyReactiveObject.WhenAnyValue(x => x.Value).Where(v => v == 2).Subscribe(_ => Debug.WriteLine("Hello World!"));
```

**Comment by RLittlesII** [comment]
@erikm-g Have you considered using [Synchronize](https://reactivex.io/documentation/operators/serialize.html) to ensure that the observables from multiple threads are received correctly?

**Comment by kmgallahan** [comment]
You may think there are cases where the lamba isn't being executed because the `Assert` shows `found` is false.

Each one of those cases is because bool isn't thread safe. Yes, when you `Assert.That(found);` the task has completed and the subscription was executed, but if the bool was set to `true` in the lambda on a different thread then it may or may not be true on the original thread at the time of the assertion - it isn't thread safe so this isn't a deterministic check.

**Comment by kmgallahan** [comment]
Also, in your original codebase initial values won't be emitted (the lamba won't be executed) if you use `WhenAnyValue(x.y.z).Subscribe` and x or y is null at the time of the subscription. That may be throwing you off as well.

**Comment by glennawatson** [comment]
We will be locking this thread soon but welcome to continue a help topic in the discussion area. Just due to the mass emails it produces

**Comment by erikm-g** [comment]
@RLittlesII I am not sure how Synchronize would help. There are no OnError or OnCompleted events and there is no issue if 2 OnNext are called at the same time. I tried adding a call to Synchronize directly after WhenAnyValue in my test and it did no difference.

@kmgallahan The test works about 299 times of 300, in those cases you will get your print out. If you check the 1 time it doesn't work it is not printed.

I think the issue is that WhenAnyValue internally gets the initial value first and starts to listen on changes after that, so any changes made in between are missed. It needs to be done the other way around to be thread safe, and if you want to avoid sending the initial value twice some check for that need to be added.

If I run this modified test where the initial value from WhenAnyValue is skipped and the current value is emitted after the subscription to changes is set up, the test succeeds every time.

```
        var anyReactiveObject = new SomeReactiveObject { Value = 1 };
        var task = Task.Run(() => anyReactiveObject.Value = 2);
        bool found = false;
        var subject = new Subject<int>();
        subject.Where(v => v == 2).Subscribe(_ => found = true);
        anyReactiveObject.WhenAnyValue(x => x.Value)
            .Skip(1).Subscribe(v => subject.OnNext(v));
        subject.OnNext(anyReactiveObject.Value);
        await task;
        Assert.That(found);
```

**Comment by kmgallahan** [comment]
@glennawatson @RLittlesII

According to the [`ConcurrentBag` docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-6.0#properties):

> All public and protected members of [ConcurrentBag<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-6.0) are thread-safe and may be used concurrently from multiple threads.

So this code should never fail, but it does sometimes:

```cs
var bag = new ConcurrentBag<int>();
var anyReactiveObject = new SomeReactiveObject();
var task = Task.Run(() => anyReactiveObject.Value = 2);

anyReactiveObject.WhenAnyValue(x => x.Value).Subscribe(x =>
{
    if (x == 2) bag.Add(2);
});

await task;

Debug.Assert(!bag.IsEmpty);
```

Either:

1. The task completes before the subscription, so a 2 should be added to the bag after `WhenAnyValue.Subscribe` since it is the initial value at that time
2. The task completes after the subscription, so the subscription should observe the property change notification and a 2 should be added to the bag

The bag should never be empty for the assertion, yet it sometimes is.

There might be a race condition inside `SubscribeToExpressionChain`:

https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI/Mixins/ReactiveNotifyPropertyChangedMixin.cs#L131

Where the sequence goes:

1. Create a `ReactiveObject` with a property that calls `RaiseAndSetIfChanged` on `set`
2. Call `Task.Run` to change the property on thread B
3. Call `WhenAnyValue.Subscribe` on the `ReactiveObject` on thread A
4. Create an `IObservable` named `notifier` loaded with an `ObservedChange` that contains the initial value of the `ReactiveObject` here:

https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI/Mixins/ReactiveNotifyPropertyChangedMixin.cs#L138

5. Thread B changes the property value on the `ReactiveObject`, calling `RaiseAndSetIfChanged` which does nothing
6. Subscribe to notifications on the property here (I think?):

https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI/Mixins/ReactiveNotifyPropertyChangedMixin.cs#L142

7. Return the `IObservable` containing the `ObservedChange` which still has Value = to the initial/unchanged value of the property as set here:

https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI/Interfaces/ObservedChange.cs#L37

In this scenario a subscription to WhenAnyValue will only emit the initial unchanged value and not the second value.

This might be resolved if property change notifications are setup before an `ObservedChange` containing the initial value is created.

I'm not entirely sure if this covers all race conditions in the logic, but it may help.

**Reply by glennawatson** [reply]
You've added indeterminacy there with the Task just firing willy nilly whenever it wants.  It could happen before or after or during the WhenAnyValue.

If you have indeterminate code you get indeterminate results and nothing to do with WhenAnyValue

**Reply by glennawatson** [reply]
Like the await is not going to defer the execution, the execution can happen whenever the task scheduler feels like it from this line 

`var task = Task.Run(() => anyReactiveObject.Value = 2);`

**Reply by erikm-g** [reply]
@glennawatson Is ReactiveUI supposed to be thread safe or not?

I don't really find anything in the documentation answering either yes or no to that question, but there are several section about scheduling and letting the logic run on any thread and then observing on a specific scheduler that strongly suggests that it is thread safe. I would find it weird to say for instance that a command can be executed on any thread and then by default the result is observed on the main thread scheduler without mentioning that the application still needs to guarantee that at most one thread at a time can interact with the command if that is an restriction.

Let's say I have an application where 2 threads add "A" and "B" respectively to a collection in parallel. If I use a regular List that is clearly stated in the documentation that it is not thread safe, I can get both "A" and "B" in the list in any order, only one of the objects in the list, only a single null value in the list, a NullReferenceException being thrown or several other outcomes. If, however, I use a thread safe collection, I could get the content {"A", "B"} or {"B", "A"} in the list, but any other outcome would be a bug in the collection and not an expected outcome of the code being indeterminate.

On the assumption that WhenAnyValue is thread safe, which the documentation lead me to believe, I can't see that the current behavior is within the expected behavior no matter the timing between the 2 threads. So I think that either this is a bug that should be fixed or the documentation needs to be updated to clearly state that this is not a thread safe library. In the latter case I wouldn't even have considered using ReactiveUI in the first place and saved myself a lot of problems.

If you have any arguments for why this is expected result from WhenAnyValue, please share, but don't continue to try to explain that Task.Run execution can happen whenever. I know that, that's the point. It is the simplest possible code to set up 2 threads running in parallel.

**Reply by glennawatson** [reply]
I would recommend that some study on task scheduling might be in order. Thread safety doesn't imply determinism which you seem to he getting confused over. I'm looking this discussion now

**Reply by glennawatson** [reply]
The examples given you have a non synchronised task around a result which is a free agent. You then after scheduling the task have a WhenAnyValue which can happen before or after the scheduled task. 

The examples I keep seeing are fundamentally broken and then constant debates about thread safety when a rogue task is scheduled without any attempted mechanisms for synchronisation. 

If you had the WhenAnyValue before the task you'd see a more.consistent result. 

Also WhenAnyValue is the same as any other Rx based mechanism and worth a read here 
http://introtorx.com/Content/v1.0.10621.0/15_SchedulingAndThreading.html#SchedulingAndThreading

**Reply by glennawatson** [reply]
The way I would think of your example with the task 

* You issue a task to run to set a value 
* You run WhenAnyValue which subscribed to the Property changed event.  Since there's no mechanism to do synchronisation with the previous free firing task it may happen before, after or during the task. 
* Therefore the values in your concurrent bag may or may not have the values you expect. 
* the await will continue execution after the task is guaranteed to have finished but the task and the WhenAnyValue have run independently of each other previous. This is the only synchronisation done in the method.

**Reply by erikm-g** [reply]
I don't expect the result to be deterministic. I expect to get first 1 then 2, or only 2, or first 2 then 1, I would even consider first 2 then 2 to be acceptable. I however think that only 1 is wrong. Just as I consider getting only 5 is wrong no matter how undeterministic the code is.

**Reply by glennawatson** [reply]
Absolutely it can be just 1. 

You don't know when in relation the subscription to INotifyPropertyChanged is going to happen with WhenAnyValue since the task is fired independently without any mechanism to link it to the WhenAnyValue

**Reply by glennawatson** [reply]
You don't have any await on the task before the WhenAnyValue nor use any mechanism to mix the task into the observable like Observable.FromTask

**Reply by erikm-g** [reply]
What is the order of events where just 1 is expected?

**Reply by glennawatson** [reply]
As per the article I linked when no scheduler is specified, also you aren't using any observable mechanism to link the task to the WhenAnyValue

By default as like any Rx in general it's a single threaded mechanism

**Reply by erikm-g** [reply]
I have read that article before and read it again. I don't see your point. I don't know the order things are happening, but don't see any order where just 1 is an expected output. If I use a BehaviorSubject with initial value 1, let the task call OnNext(2) and subscribing to the BehaviorSubject instead of using WhenAnyValue I never just 1. I always get at least a 2.

**Reply by erikm-g** [reply]
The task is actually a callback from a MQTT communication library setting the value of a property in the model if it is connected or not, the WhenAnyValue and subscription is made in a VM. I don't know how they should be linked in any other way than that one sets the property and the other is listening to changes of it.

**Reply by glennawatson** [reply]
Can you move the WhenAnyValue before any potential value being set. That's likely to eliminate your issue

**Reply by erikm-g** [reply]
No, in this case I can't do that. In most cases they are set up in the constructors during program initialization and then I don't run into any issue, but in a few cases the subscription needs to be started later.

**Reply by glennawatson** [reply]
Getting late here so will have to look more tomorrow. Likely a fix could be provided once off in a ReactiveObject for your scenario with the property and the trigger for INotifyPropertyChanged being locked ()

**Comment by anaisbetts** [comment, marked as answer]
WhenAny is not "thread-safe", it has the same thread safety rules as any other C# object. You should not write to ViewModel properties on other threads, as they inherit the thread affinity of the Views they are bound to. If you wouldn't write from other threads to a View, you shouldn't do it to a ViewModel!

**Reply by kmgallahan** [reply]
> You should not write to ViewModel properties on other threads, as they inherit the thread affinity of the Views they are bound to. If you wouldn't write from other threads to a View, you shouldn't do it to a ViewModel!

We can all agree on this:

* This is an edge case that relies on a narrow timing sequence and may only affect 1 person in production
* The provided code samples look terrible and can easily be changed to avoid the edge case

And I think we can all agree on this:

* It is (common practice / not unheard of / valid)? to use WhenAnyValue to observe properties that can be modified by other threads, as subscriptions can be handled on the UI thread (or it can be used in contexts that never involve a UI thread)
* Complex multithreaded environments exist where hitting this edge case isn't so obvious, nor mitigation so easy

So the question is:

* Is there a simple code change that could eliminate or reduce occurrences of the edge case without negative side effects? 

Such as using a lock or setting up property notifications before getting the initial value [here](https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI/Mixins/ReactiveNotifyPropertyChangedMixin.cs#L131):

```cs
IObservable<IObservedChange<object?, object?>> notifier =
            Observable.Return(new ObservedChange<object?, object?>(null, null, source));

        var chain = Reflection.Rewrite(expression).GetExpressionChain();
        notifier = chain.Aggregate(notifier, (n, expr) => n
                                                          .Select(y => NestedObservedChanges(expr, y, beforeChange, suppressWarnings))
                                                          .Switch());

        if (skipInitial)
        {
            notifier = notifier.Skip(1);
        }
```

I fully acknowledge that I don't know all the intricacies of the code, and I may be thinking in an incredibly naive manner here.

I apologize for dragging this conversation out. This scenario just bothers me a bit:

1. `Value` = "Hello"
2. Do an arbitrary amount of work for an hour
3. Subscribe to `WhenAnyValue` for `Value`
4. Do an arbitrary amount of work an hour

If you set `Value` to "World!" on a different thread (even with a lock) anytime during those two hours then your code can be in this state:

`Value` = "World!" | `WhenAnyValue` has never emitted "World!"

**Reply by anaisbetts** [reply]
> It is (common practice / not unheard of / valid)? to use WhenAnyValue to observe properties that can be modified by other threads, as subscriptions can be handled on the UI thread

I do _not_ agree with this, if you're doing this, you are making a Mistake, full-stop. While your example with the ConcurrentBag is definitely Curious (I tend to agree that the current code should _theoretically_ work and I don't see why it wouldn't immediately), it is also Invalid - you _shouldn't_ do this. What is valid:

1. Creating a ViewModel on another thread, initializing it, then handing it to the UI thread (good for loading datasets etc) to be bound to the UI.
2. Accessing ViewModels on the UI thread

The easiest way to fix these problems is not to try to fix your class, but to instead fix the setter - walk up the stack until you find the code setting the value and wrap it in `RxApp.MainThreadScheduler.Schedule(()=> { ... });`. Note that this isn't always the most _efficient_ way to fix it though!

**Reply by kmgallahan** [reply]
> the ConcurrentBag is definitely Curious (I tend to agree that the current code should theoretically work

The reason the code can fail is that this timing sequence is possible:

1. WhenAnyValue gets the initial value
2. Other thread changes value
3. WhenAnyValue sets up property notifications for value

As an outsider it seems like a library level change could make this edge case go away for every ReactiveUI user, rather than having most users not be aware it can happen or have to put guards for it throughout their code.

I also find WhenAnyValue incredibly useful, so much so that relegating it to only ViewModel related tasks seems like a shame (even though apparently this is the only way it was intended to be used).

>  wrap it in RxApp.MainThreadScheduler.Schedule

Unfortunately Reactive.NET and ReactiveUI don't support [WinUI 3's new DispatcherQueue](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.dispatching.dispatcherqueue?view=winui-3.0). ~Of course it can still be used via other methods, or devs can use a lock on a property when sending it through WhenAnyValue.

[#1651](https://github.com/dotnet/reactive/issues/1651)

**Reply by glennawatson** [reply]
It has supported the dispatcherqueue for a while now 

https://github.com/reactiveui/ReactiveUI/blob/86283536c38a84e2f5a4b92e69076903c25b551a/src/ReactiveUI.WinUI/DispatcherQueueScheduler.cs#L12

**Reply by kmgallahan** [reply]
@glennawatson Oh wow nice, thanks! I've been scanning the release notes and missed this. Also thought ObserveOn would be involved in some way.

**Reply by anaisbetts** [reply]
> As an outsider it seems like a library level change could make this edge case go away for every ReactiveUI user, rather than having most users not be aware it can happen or have to put guards for it throughout their code

In practice, I suspect this doesn't hit as often as you'd think, because WhenAnyValue should almost exclusively be called in the object's constructor. That being said, I would support moving the initial value fetch to be later but I would not try to fix this in any more invasive way.

> Unfortunately Reactive.NET and ReactiveUI don't support [WinUI 3's new DispatcherQueue](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.dispatching.dispatcherqueue?view=winui-3.0).

Nearly every platform facet of ReactiveUI is replaceable, and while @glennawatson mentioned that this is already done, you can actually do it in your own project too without making changes to ReactiveUI itself. You can make your own `IScheduler` implementation (it's very easy, only a few methods), then all you have to do is:

```cs
RxApp.MainThreadSchedule = new MyCoolScheduler();
```

You can actually code 100% support for an entirely new UI framework without changing a line of code in ReactiveUI, by registering new implementations of interfaces like `IActivationForViewFetcher`, `ICreatesObservableForProperty` and `IBindingTypeConverter`

**Reply by anaisbetts** [reply]
> Also thought ObserveOn would be involved in some way.

`ObserveOn` is basically just calling `Schedule` for every item in an Observable

**Reply by glennawatson** [reply]
Btw don't think at this stage there enough need to warrant changing the fetch to later so I won't be making that change. The issue was moved to a discussion since there are already mechanisms in place to get everything onto the main thread you can already use. Eg use Observable.FromAsync to get the result back onto the main thread scheduled.

**Reply by erikm-g** [reply]
> > It is (common practice / not unheard of / valid)? to use WhenAnyValue to observe properties that can be modified by other threads, as subscriptions can be handled on the UI thread
> 
> I do _not_ agree with this, if you're doing this, you are making a Mistake, full-stop. 


If so, why is this code example:
```
this.WhenAnyValue(x => x.MyImportantProperty).ObserveOn(RxApp.MainThreadScheduler).Subscribe(x => ...);
```

listed on https://www.reactiveui.net/docs/handbook/scheduling/ ?

**Reply by glennawatson** [reply]
If you use it in the constructor you can handle properties with different threads being set.

**Reply by glennawatson** [reply]
But multithreaded property setting is not recommended. The documentation could use improvement.

**Reply by erikm-g** [reply]
That is my point. If this is not a bug that should be corrected, the documentation needs to be updated. I would of course have preferred if the code was updated to handle multithreading as I thought it would after reading the documentation, but I can understand if that is not feasible.

**Reply by anaisbetts** [reply]
> If so, why is this code example listed on https://www.reactiveui.net/docs/handbook/scheduling/

That is a bad code example and we should for sure yank it, thanks for pointing this out @erikm-g
